### PR TITLE
UI for adding Attachments when composing new Message

### PR DESCRIPTION
**Pull request by LunarX**
Depends on https://github.com/Infomaniak/android-core/pull/21

### CODE REVIEW

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/AttachmentAdapter.kt [hunk 1]
```diff
@@ -41,16 +44,36 @@ class AttachmentAdapter(
         val attachment = attachments[position]
 
         fileName.text = attachment.name
-        // TODO: Find how to add the FileType prefix before the file' size.
-        fileDetails.text = /*item.mimeType + " - " + */FormatterFileSize.formatShortFileSize(context, attachment.size.toLong())
+        fileDetails.text = FormatterFileSize.formatShortFileSize(context, attachment.size.toLong())
         icon.load(attachment.getFileTypeFromExtension().icon)
```
Comment (JorisBodin) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/AttachmentAdapter.kt:48]: You must do as in kDrive, display the file thumbnail directly from the local storage.
https://github.com/Infomaniak/android-kDrive/blob/master/app/src/main/java/com/infomaniak/drive/utils/FileItemUtils.kt#L131

File: app/src/main/java/com/infomaniak/mail/utils/LocalStorageUtils.kt [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        if (userDir.hasEmptyFiles) userDir.delete()
+        if (attachmentsRootDir.hasEmptyFiles) attachmentsRootDir.delete()
+    }
+
+    fun deleteUserData(context: Context, userId: Int) {
+        with(context.attachmentsRootDir) {
+            File(this, "$userId").deleteRecursively()
+            if (this.listFiles()?.isEmpty() == true) deleteRecursively()
+        }
+    }
```
Comment (JorisBodin) [inline app/src/main/java/com/infomaniak/mail/utils/LocalStorageUtils.kt:95]: `userId: Int) = with(context.` or a extention of `File`

File: app/src/main/java/com/infomaniak/mail/utils/Extensions.kt [hunk 3]
```diff
@@ -140,6 +143,25 @@ fun Fragment.safeNavigateToNewMessageActivity(draftMode: DraftMode, messageUid:
...
+                Sentry.captureException(Exception("$this has empty cursor"))
+                null
+            }
+        }
+    }.getOrElse { exception ->
+        Sentry.captureException(exception)
+        null
+    }
+}
+
```
Comment (JorisBodin) [inline app/src/main/java/com/infomaniak/mail/utils/Extensions.kt:164]: From experience with kDrive, the `DISPLAY_NAME` and `SIZE` can be null depending on the phones and applications that created them.
It's a bad thing to rewrite code, which could be shared with kDrive in Core
https://github.com/Infomaniak/android-kDrive/blob/43357f5e2a9fbe2c47ea644ec9d36a92067fa15c/app/src/main/java/com/infomaniak/drive/utils/SyncUtils.kt#L47

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/AttachmentAdapter.kt [hunk 4]
```diff
@@ -41,16 +44,36 @@ class AttachmentAdapter(
...
-        root.setOnClickListener { onAttachmentClicked?.invoke(attachment) }
+
+        if (!shouldDisplayCloseButton) {
+            root.setOnClickListener { onAttachmentClicked?.invoke(attachment) }
+        } else {
+            closeButton.apply {
+                setOnClickListener {
+                    val index = attachments.indexOf(attachment)
+                    if (index != -1) {
```
Comment (JorisBodin) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/AttachmentAdapter.kt:56]: When can it be -1?

File: app/src/main/java/com/infomaniak/mail/ui/main/newMessage/NewMessageViewModel.kt [hunk 5]
```diff
@@ -150,21 +154,58 @@ class NewMessageViewModel(application: Application) : AndroidViewModel(applicati
...
+                newAttachments.add(it)
+                attachmentsSize += it.size
+            }
+        }
+
+        importedAttachments.postValue(newAttachments to ImportationResult.SUCCESS)
+    }
+
+    private fun importAttachment(uri: Uri, availableSpace: Int): Pair<Attachment?, Boolean> {
+        val (fileName, fileSize) = uri.getFileNameAndSize(getApplication())!!
```
Comment (JorisBodin) [inline app/src/main/java/com/infomaniak/mail/ui/main/newMessage/NewMessageViewModel.kt:185]: No `!!' in this situation

File: app/src/main/java/com/infomaniak/mail/ui/main/newMessage/NewMessageViewModel.kt [hunk 6]
```diff
@@ -150,21 +154,58 @@ class NewMessageViewModel(application: Application) : AndroidViewModel(applicati
...
+        if (fileSize > availableSpace) return null to true
+
+        return LocalStorageUtils.copyDataToAttachmentsCache(getApplication(), uri, fileName, currentDraftLocalUuid)?.let { file ->
+            val fileExtension = file.path.substringAfterLast(".")
+            val mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(fileExtension) ?: "*/*"
```
Comment (JorisBodin) [inline app/src/main/java/com/infomaniak/mail/ui/main/newMessage/NewMessageViewModel.kt:190]: A method exists in kDrive, so put in Core at this time
https://github.com/Infomaniak/android-kDrive/blob/d56ee5599005ce97521d79207d3522e2839d1638/app/src/main/java/com/infomaniak/drive/data/models/File.kt#L307

**Review by JorisBodin — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/AttachmentAdapter.kt (hunk 4 above)
Comment (LunarX) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/AttachmentAdapter.kt:56]: When clicking on the the attachment in order to delete it, we remove the attachment from the list successfully which means that during the fade out animation, if we click again on the little cross the attachment is not in the list anymore yet the cross has been clicked. This results in a negative index which is why this check is for

Comment (JorisBodin) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/AttachmentAdapter.kt:56]: Can you put a comment in the code to that end please?

**Review by sirambd — APPROVED** [review]

**Review by JorisBodin — APPROVED** [review]

**Pull request merged**